File: src/main/java/com/extendedae_plus/mixin/impl/ProviderSettingsImplementations.java
package com.extendedae_plus.mixin.impl;

import appeng.api.config.Setting;
import appeng.api.config.Settings;
import appeng.api.config.YesNo;
import appeng.api.crafting.IPatternDetails;
import appeng.api.stacks.AEKey;
import appeng.api.util.IConfigManager;
import appeng.helpers.patternprovider.PatternProviderTarget;
import com.extendedae_plus.common.impl.pattern.smartDoubling.SmartDoublingAwarePattern;
import com.extendedae_plus.common.init.ModSettings;
import com.extendedae_plus.common.settings.StateSmartBlocking;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public final class ProviderSettingsImplementations {
    private static final List<IConfigManager> flagChanging = new ArrayList<>();

    public static void onSettingsChanged(IConfigManager manager,
                                         Setting<?> setting,
                                         Runnable doublingUpdater) {
        if (flagChanging.contains(manager)) return;

        if (Settings.BLOCKING_MODE.equals(setting)) {
            var stateBlocking = switch (manager.getSetting(Settings.BLOCKING_MODE)) {
                case NO -> StateSmartBlocking.DISABLED_BY_SUPER;
                case YES -> StateSmartBlocking.DISABLED;
                default -> null;
            };
            if (stateBlocking != null) {
                flagChanging.add(manager);
                manager.putSetting(ModSettings.SMART_BLOCKING, stateBlocking);
                flagChanging.remove(manager);
            }
        } else if (ModSettings.SMART_BLOCKING.equals(setting)
                && StateSmartBlocking.ENABLED.equals(manager.getSetting(setting))) {
            flagChanging.add(manager);
            manager.putSetting(Settings.BLOCKING_MODE, YesNo.YES);
            flagChanging.remove(manager);
        } else if (ModSettings.SMART_DOUBLING.equals(setting)) {
            doublingUpdater.run();
        }
    }

    public static boolean checkCanBlock(boolean isBlocking,
                                        IConfigManager configManager,
                                        PatternProviderTarget target,
                                        Set<AEKey> inputs,
                                        IPatternDetails patternDetails) {
        return target.containsPatternInput(inputs)
                && (!isBlocking
                || (!StateSmartBlocking.ENABLED.equals(configManager.getSetting(ModSettings.SMART_BLOCKING))
                || !matchBlockingInputs(target, patternDetails)));
    }

    public static void updateDoublingState(IConfigManager configManager, List<IPatternDetails> patterns) {
        patterns.forEach(pattern -> {
            if (!(pattern instanceof SmartDoublingAwarePattern doublingPattern)) return;
            doublingPattern.eap$setAllowScaling(
                    YesNo.YES.equals(configManager.getSetting(ModSettings.SMART_DOUBLING)));
        });
    }

    private static boolean matchBlockingInputs(PatternProviderTarget target, IPatternDetails patternDetails) {
        for (var slot : patternDetails.getInputs()) {
            var inputs = Arrays.stream(slot.getPossibleInputs())
                    .map(input -> input.what().dropSecondary())
                    .collect(Collectors.toSet());
            if (!target.containsPatternInput(inputs)) return false;
        }
        return true;
    }
}

File: src/main/kotlin/com/extendedae_plus/mixin/impl/ProviderSettingsImplementations.kt
package com.extendedae_plus.mixin.impl

import appeng.api.config.Setting
import appeng.api.config.Settings
import appeng.api.config.YesNo
import appeng.api.crafting.IPatternDetails
import appeng.api.stacks.AEKey
import appeng.api.util.IConfigManager
import appeng.helpers.patternprovider.PatternProviderTarget
import com.extendedae_plus.common.init.EAEPSettings
import com.extendedae_plus.common.registry.settings.StateSmartBlocking
import com.extendedae_plus.util.extension.ExtensionScaledPattern

object ProviderSettingsImplementations {
    @JvmField
    val flagChanging = ArrayList<IConfigManager>()

    @JvmStatic
    fun onSettingsChanged(
        manager: IConfigManager,
        setting: Setting<*>,
        doublingUpdater: Runnable
    ) {
        if (flagChanging.contains(manager)) return

        if (Settings.BLOCKING_MODE == setting) {
            val stateBlocking = when (manager.getSetting(Settings.BLOCKING_MODE)) {
                YesNo.NO -> StateSmartBlocking.DISABLED_BY_SUPER
                YesNo.YES -> StateSmartBlocking.DISABLED
                else -> null
            }
            if (stateBlocking != null) {
                flagChanging.add(manager)
                manager.putSetting(EAEPSettings.smartBlocking, stateBlocking)
                flagChanging.remove(manager)
            }
        } else if (EAEPSettings.smartBlocking == setting
            && StateSmartBlocking.ENABLED == manager.getSetting(setting)
        ) {
            flagChanging.add(manager)
            manager.putSetting(Settings.BLOCKING_MODE, YesNo.YES)
            flagChanging.remove(manager)
        } else if (EAEPSettings.smartDoubling == setting) {
            doublingUpdater.run()
        }
    }

    @JvmStatic
    fun checkCanBlock(
        isBlocking: Boolean,
        configManager: IConfigManager,
        target: PatternProviderTarget,
        inputs: Set<AEKey>,
        patternDetails: IPatternDetails
    ) = target.containsPatternInput(inputs)
            && (!isBlocking
            || (StateSmartBlocking.ENABLED != configManager.getSetting(EAEPSettings.smartBlocking)
            || !matchBlockingInputs(target, patternDetails)))

    @JvmStatic
    fun updateDoublingState(configManager: IConfigManager, patterns: List<IPatternDetails>) =
        patterns.forEach(ExtensionScaledPattern.setState(configManager.getSetting(EAEPSettings.smartDoubling)))

    private fun matchBlockingInputs(target: PatternProviderTarget, patternDetails: IPatternDetails): Boolean {
        for (slot in patternDetails.inputs) {
            val inputs = slot.possibleInputs
                .map { it.what().dropSecondary() }
                .toSet()
            if (!target.containsPatternInput(inputs)) return false
        }
        return true
    }
}
